source/common/router/header_parser.cc

```diff
@@ -1,28 +1,206 @@
+  ExpectVariableEnd          // expect closing % in %VAR(...)%
+};
+
+std::string unescape(absl::string_view sv) { return absl::StrJoin(absl::StrSplit(sv, "%%"), "%"); }
+
+// Implements a state machine to parse custom headers. Each character of the custom header format
+// is either literal text (with % escaped as %%) or part of a %VAR% or %VAR(["args"])% expression.
+// The statement machine does minimal validation of the arguments (if any) and does not know the
+// names of valid variables. Interpretation of the variable name and arguments is delegated to
+// RequestInfoHeaderFormatter.
 HeaderFormatterPtr parseInternal(const envoy::api::v2::HeaderValueOption& header_value_option) {
-  const std::string& format = header_value_option.header().value();
   const bool append = PROTOBUF_GET_WRAPPED_OR_DEFAULT(header_value_option, append, true);
 
-  if (format.find("%") == 0) {
-    const size_t last_occ_pos = format.rfind("%");
-    if (last_occ_pos == std::string::npos || last_occ_pos <= 1) {
-      throw EnvoyException(fmt::format("Incorrect header configuration. Expected variable format "
-                                       "%<variable_name>%, actual format {}",
-                                       format));
+  absl::string_view format(header_value_option.header().value());
+  if (format.empty()) {
+    return HeaderFormatterPtr{new PlainHeaderFormatter("", append)};
+  }
+
+  std::vector<HeaderFormatterPtr> formatters;
+
+  size_t pos = 0, start = 0;
+  ParserState state = ParserState::Literal;
+  do {
+    const char ch = format[pos];
+    const bool has_next_ch = (pos + 1) < format.size();
+
+    switch (state) {
+    case ParserState::Literal:
+      // Searching for start of %VARIABLE% expression.
+      if (ch != '%') {
+        break;
+      }
+
+      if (!has_next_ch) {
+        throw EnvoyException(
+            fmt::format("Invalid header configuration. Un-escaped % at position {}", pos));
+      }
+
+      if (format[pos + 1] == '%') {
+        // Escaped %, skip next character.
+        pos++;
+        break;
+      }
+
+      // Un-escaped %: start of variable name. Create a formatter for preceding characters, if
+      // any.
+      state = ParserState::VariableName;
+      if (pos > start) {
+        absl::string_view literal = format.substr(start, pos - start);
+        formatters.emplace_back(new PlainHeaderFormatter(unescape(literal), append));
+      }
+      start = pos + 1;
+      break;
+
+    case ParserState::VariableName:
+      // Consume "VAR" from "%VAR%" or "%VAR(...)%"
+      if (ch == '%') {
+        // Found complete variable name, add formatter.
+        formatters.emplace_back(
+            new RequestInfoHeaderFormatter(format.substr(start, pos - start), append));
+        start = pos + 1;
+        state = ParserState::Literal;
+        break;
+      }
+
+      if (ch == '(') {
+        // Variable with arguments, search for start of arg array.
+        state = ParserState::ExpectArray;
```

Comment:
I was expecting you to bump a depth-count here.  Scanning forward a bit, it looks like the algorithm is very sensitive to a particular json structure.  I think it might actually be just as easy (and maybe very few if any additional lines of code) to have your scanner track the paren depth for ```() [] {}``` and the quoting state for `"` and `'`.  In a way it might be easier to review as the reviewer would just need to take into account basic lexical nesting and not the schema you are tuned to.

Perhaps this is speculative generality but I also wonder if the json structure you expect might change over time requiring the next person to then update this routine.

Reply:
The history of using JSON here is that when custom headers only supported the entire header value being custom (vs. what this PR is doing) we thought it would be easier to just use the JSON parser to handle arguments rather than writing an argument parser. In retrospect, I wish I had just handled arguments. FWIW, I'd be willing to toss the JSON business and rewrite this parser to just take quoted arguments (e.g. `%UPSTREAM_METADATA("a", "b", "c")%`. But if we keep JSON, the upshot is that I don't expect anyone to want anything other than an array. Something more complicated would (IMO) warrant modifying the protobuf definition of HeaderValueOption.

The [JSON spec](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf) doesn't seem to allow single-quoted strings, so I don't think we need to handle them in any event.

Reply:
This comment spawned a bunch of thought from me, but not that much typing.  As I'm relatively new to Envoy I still don't have a great sense of what needs to be fast (e.g. runs per request) and what can't be changed (because it's already released and documented).

If this runs per request, I suspect it'd be worth considering dumping JSON parsing and just scan using string_view-related routines.  If this only runs at startup, I'm a little less concerned, however I'm also looking at https://github.com/istio/istio/issues/2554#issuecomment-358192433 and https://github.com/envoyproxy/envoy/issues/2373 which are all around startup speed.  That doesn't mean *this* would be a problem but I think it's worth understanding it.

The second part of this: is the syntax you need to parse set in stone?  If so, and JSON generality is not needed, it'd be nice to drop it in favor of something that could be more easily and quickly parsed.

For the moment I'll continue the review assuming the answers are "the speed does not matter" and "the format cannot be changed".

Reply:
I don't have a strong opinion. This seems fine to me. What's the proposed alternative?

Reply:
Right now the parser does just enough parsing to find the start/end of % expressions, but needs to know something about JSON to do it. We could just make the parser smart enough to parse the params and ditch the JSON parser. While we're there we could drop the square brackets. If you want to do this I'll put in another PR.